File: sdks/python/superagent/streaming.py
"""A2UI SSE streaming support."""

from __future__ import annotations

import asyncio
import json
from typing import Any, AsyncIterator, Callable, Coroutine, Dict, List, Optional

import httpx

from .exceptions import StreamDisconnectedError
from .types import A2UIEvent, A2UIEventType


class A2UIStream:
    """Async iterator over A2UI Server-Sent Events.

    Usage::

        stream = client.chat_stream("my-agent", "Hello")
        async for event in stream:
            if event.event_type == "text":
                print(event.text_delta, end="", flush=True)

    Alternatively use the high-level helpers::

        text = await stream.collect()
    """

    def __init__(self, response: httpx.Response) -> None:
        self._response = response
        self._iter: AsyncIterator[str] = response.aiter_lines()
        self._done = False
        # pending SSE fields before the blank-line flush
        self._pending_event: Optional[str] = None
        self._pending_data: Optional[str] = None
        # registered callbacks: event_type -> list[callback]
        self._callbacks: Dict[str, List[Callable[[A2UIEvent], Any]]] = {}

    # ------------------------------------------------------------------
    # Async iterator protocol
    # ------------------------------------------------------------------

    def __aiter__(self) -> "A2UIStream":
        return self

    async def __anext__(self) -> A2UIEvent:
        """Return the next parsed A2UI event, or raise StopAsyncIteration."""
        if self._done:
            raise StopAsyncIteration

        try:
            return await self._next_event()
        except StopAsyncIteration:
            self._done = True
            raise
        except httpx.RemoteProtocolError as exc:
            self._done = True
            raise StreamDisconnectedError(str(exc)) from exc

    async def _next_event(self) -> A2UIEvent:
        """Read SSE lines until a complete event is assembled."""
        event_type: Optional[str] = None
        data_lines: List[str] = []

        async for line in self._iter:
            line = line.rstrip("\r")

            if line == "":
                # Blank line: dispatch accumulated event if any
                if data_lines:
                    raw_data = "\n".join(data_lines)
                    evt = self._parse_event(event_type or "message", raw_data)
                    event_type = None
                    data_lines = []
                    await self._fire_callbacks(evt)
                    # Stop iteration after done event
                    if evt.event_type == A2UIEventType.done:
                        self._done = True
                    return evt
                # Empty flush — keep reading
                continue

            if line.startswith("event:"):
                event_type = line[6:].strip()
            elif line.startswith("data:"):
                data_lines.append(line[5:].strip())
            # Ignore id:, retry:, and comment lines

        raise StopAsyncIteration

    @staticmethod
    def _parse_event(event_type: str, raw_data: str) -> A2UIEvent:
        """Parse a raw SSE data payload into an A2UIEvent."""
        try:
            parsed = json.loads(raw_data)
            if isinstance(parsed, dict):
                # Server may embed event type inside the payload
                if "type" in parsed and event_type == "message":
                    event_type = parsed.pop("type")
                data: Dict[str, Any] = parsed
            else:
                data = {"value": parsed}
        except (json.JSONDecodeError, ValueError):
            data = {"raw": raw_data}

        return A2UIEvent(event_type=event_type, data=data)

    # ------------------------------------------------------------------
    # High-level helpers
    # ------------------------------------------------------------------

    async def collect(self) -> str:
        """Consume the entire stream and return concatenated text content."""
        parts: List[str] = []
        async for event in self:
            if event.event_type in (A2UIEventType.text, "text"):
                delta = event.data.get("delta", "") or event.data.get("content", "")
                if delta:
                    parts.append(delta)
        return "".join(parts)

    def on(
        self,
        event_type: str | A2UIEventType,
        callback: Callable[[A2UIEvent], Any],
    ) -> "A2UIStream":
        """Register a callback for a specific event type.

        The callback receives the :class:`A2UIEvent` and may be a plain
        function or a coroutine function.  Returns *self* for chaining.

        Example::

            stream.on("tool_call", lambda e: print("tool:", e.data))
        """
        key = event_type.value if isinstance(event_type, A2UIEventType) else event_type
        self._callbacks.setdefault(key, []).append(callback)
        return self

    async def _fire_callbacks(self, event: A2UIEvent) -> None:
        handlers = self._callbacks.get(event.event_type, [])
        for handler in handlers:
            result = handler(event)
            if asyncio.iscoroutine(result):
                await result

    async def aclose(self) -> None:
        """Close the underlying HTTP response."""
        await self._response.aclose()

    async def __aenter__(self) -> "A2UIStream":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.aclose()

File: sdks/typescript/src/streaming.ts
import { A2UIEvent, A2UIEventType } from "./types.js";
import { StreamDisconnectedError } from "./errors.js";

type EventHandler = (data: Record<string, unknown>) => void;

/**
 * A2UIStream wraps a streaming HTTP response and exposes A2UI events via
 * two consumption patterns:
 *
 *   1. Async iterator: `for await (const event of stream) { ... }`
 *   2. Event listener: `stream.on('text', handler).on('done', handler)`
 *
 * Only one consumption pattern should be used per stream instance.
 */
export class A2UIStream implements AsyncIterable<A2UIEvent> {
  private readonly responsePromise: Promise<Response>;
  private readonly controller: AbortController;
  private readonly callbacks: Map<string, EventHandler[]> = new Map();

  constructor(responsePromise: Promise<Response>, controller: AbortController) {
    this.responsePromise = responsePromise;
    this.controller = controller;
  }

  // ---------------------------------------------------------------------------
  // Event listener API
  // ---------------------------------------------------------------------------

  /**
   * Register a handler for a specific A2UI event type.
   * Returns `this` to allow chaining.
   *
   * @example
   * stream
   *   .on(A2UIEventType.Text, ({ delta }) => process.stdout.write(String(delta)))
   *   .on('done', () => console.log('finished'));
   */
  on(
    event: A2UIEventType | "text" | "error" | "done",
    handler: EventHandler,
  ): this {
    const list = this.callbacks.get(event) ?? [];
    list.push(handler);
    this.callbacks.set(event, list);
    return this;
  }

  // ---------------------------------------------------------------------------
  // Async iterator API
  // ---------------------------------------------------------------------------

  [Symbol.asyncIterator](): AsyncIterator<A2UIEvent> {
    return this._iterate();
  }

  private async *_iterate(): AsyncGenerator<A2UIEvent> {
    const response = await this.responsePromise;
    if (!response.ok) {
      await this._throwHttpError(response);
    }
    if (!response.body) {
      throw new StreamDisconnectedError("Response body is null");
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const frames = buffer.split("\n\n");
        // Keep the last (potentially incomplete) frame in the buffer.
        buffer = frames.pop() ?? "";

        for (const frame of frames) {
          const event = parseSSEFrame(frame);
          if (event !== null) {
            this._dispatch(event);
            yield event;
          }
        }
      }

      // Flush any remaining buffer content.
      if (buffer.trim()) {
        const event = parseSSEFrame(buffer);
        if (event !== null) {
          this._dispatch(event);
          yield event;
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  // ---------------------------------------------------------------------------
  // Convenience: consume all events and return concatenated text
  // ---------------------------------------------------------------------------

  /**
   * Iterates the stream to completion and returns all text delta content
   * concatenated into a single string.
   */
  async collect(): Promise<string> {
    let result = "";
    for await (const event of this) {
      if (event.eventType === A2UIEventType.Text) {
        const delta = event.data["delta"];
        if (typeof delta === "string") {
          result += delta;
        }
      }
    }
    return result;
  }

  // ---------------------------------------------------------------------------
  // Cancel
  // ---------------------------------------------------------------------------

  /** Abort the underlying HTTP request and close the stream. */
  abort(): void {
    this.controller.abort();
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  /**
   * Start consuming the stream using the event listener callbacks.
   * Called automatically when `.on()` is used before iteration begins.
   */
  async _startCallbackLoop(): Promise<void> {
    try {
      for await (const event of this) {
        // Events are dispatched inside _iterate() via _dispatch().
        // Nothing extra needed here; the loop just drives consumption.
        void event;
      }
    } catch (err) {
      const handlers = this.callbacks.get("error") ?? [];
      const payload =
        err instanceof Error
          ? { message: err.message, code: "stream_error" }
          : { message: String(err), code: "stream_error" };
      for (const h of handlers) {
        h(payload);
      }
    }
  }

  private _dispatch(event: A2UIEvent): void {
    const handlers = this.callbacks.get(event.eventType) ?? [];
    for (const h of handlers) {
      h(event.data);
    }
  }

  private async _throwHttpError(response: Response): Promise<never> {
    let message = `HTTP ${response.status}`;
    try {
      const body = await response.text();
      const json = JSON.parse(body) as Record<string, unknown>;
      if (typeof json["msg"] === "string") message = json["msg"];
      else if (typeof json["error"] === "string") message = json["error"];
      else if (typeof json["message"] === "string") message = json["message"];
    } catch {
      // Ignore parse errors; keep the default message.
    }

    const { SuperagentError, AuthenticationError, InterruptConflictError } =
      await import("./errors.js");
    if (response.status === 401 || response.status === 403) {
      throw new AuthenticationError(message);
    }
    if (response.status === 409) {
      throw new InterruptConflictError(message);
    }
    throw new SuperagentError(response.status, "http_error", message);
  }
}

// ---------------------------------------------------------------------------
// SSE frame parser
// ---------------------------------------------------------------------------

/**
 * Parse a single SSE frame (the text between two `\n\n` separators).
 *
 * Expected wire format from the Superagent backend (A2UI mode):
 *   event: <A2UIEventType>
 *   data: <JSON string>
 *
 * Returns null for comment-only or empty frames.
 */
function parseSSEFrame(frame: string): A2UIEvent | null {
  let eventType: string | null = null;
  let dataLine: string | null = null;

  for (const line of frame.split("\n")) {
    if (line.startsWith("event:")) {
      eventType = line.slice(6).trim();
    } else if (line.startsWith("data:")) {
      dataLine = line.slice(5).trim();
    }
    // Ignore id:, retry:, and comment lines (:).
  }

  if (!dataLine) return null;

  // Handle legacy plain-token mode where the backend emits `data: <token>`.
  if (eventType === null || eventType === "message") {
    if (dataLine === "[DONE]") {
      return {
        eventType: A2UIEventType.Done,
        data: {},
      };
    }
    return {
      eventType: A2UIEventType.Text,
      data: { delta: dataLine, content: dataLine },
    };
  }

  let parsed: Record<string, unknown> = {};
  try {
    const raw = JSON.parse(dataLine) as unknown;
    // The backend wraps the payload in an Event envelope: { type, timestamp, data }.
    // Extract the inner data field when present.
    if (
      raw !== null &&
      typeof raw === "object" &&
      "data" in (raw as object) &&
      typeof (raw as Record<string, unknown>)["data"] === "object" &&
      (raw as Record<string, unknown>)["data"] !== null
    ) {
      parsed = (raw as Record<string, unknown>)["data"] as Record<
        string,
        unknown
      >;
    } else if (raw !== null && typeof raw === "object") {
      parsed = raw as Record<string, unknown>;
    }
  } catch {
    // Non-JSON data; wrap as raw content.
    parsed = { raw: dataLine };
  }

  return {
    eventType: eventType as A2UIEventType,
    data: parsed,
  };
}
